External event capture for photogrammetry etc.

henryhallam opened the pull request:
There's a new setting, ext_events.edge_trigger, that selects which (if any)
edges on DEBUG0 will trigger a timestamped event report.

Presently, events are output as an INFO message.  A new SBP message type is
being defined to replace that.

henryhallam commented:
Code review: @fnoble @cbeighley @soulcmdc @gsmcmullin 

Todo:
* Waiting on SBP stability to use MSG_EXT_EVENT rather than an info message
* Convert to GPS time.

swiftnav-jenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://zazu.ci.swiftnav.com/job/piksi_firmware_Pull_Requests/249/
Test FAILed.

swiftnav-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://zazu.ci.swiftnav.com/job/piksi_firmware_Pull_Requests/250/
Test PASSed.

swiftnav-jenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://zazu.ci.swiftnav.com/job/piksi_firmware_Pull_Requests/252/
Test FAILed.

henryhallam commented:
This is working nicely, I'm happy with the structure and it's ready for review. @fnoble @cbeighley @denniszollo

swiftnav-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://zazu.ci.swiftnav.com/job/piksi_firmware_Pull_Requests/253/
Test PASSed.

mfine commented:
Heads up! #354 is going to remove the `scripts/` directory and this commit contains `scripts/` changes. Before removing, let's get these changes in and brought over to `piksi_tools`.

fnoble commented on src/board/nap/nap_common.h:
```diff
@@ -66,6 +66,17 @@ void nap_conf_b_clear(void);
+ * \param reg_id   NAP register ID.
+ * \return u32 register value
+ */
+inline u32 nap_read_u32(u8 reg_id) {
```
`{` on its own line

fnoble commented on src/board/nap/nap_ext_event.c:
```diff
@@ -0,0 +1,97 @@
+  {
+    u8 v[5]={0};
+    v[4] = trigger;
+    nap_xfer_blocking(NAP_REG_EXT_EVENT_TIME, 5, v, v);
```
Maybe make a function to set this register and move to `nap_common`

fnoble commented on src/board/nap/nap_ext_event.c:
```diff
@@ -0,0 +1,97 @@
+ * \{ */
+
+typedef enum {
+  NONE,
```
If you rely on the numerical value of these then you should explicitly define them

fnoble commented on src/board/nap/nap_ext_event.c:
```diff
@@ -0,0 +1,97 @@
+    } d;
+    u8 b[5];
+  } v;
+  v.b[4] = trigger;  // We have to reset the trigger for next time
```
change all `//` to `/* */`

fnoble commented on src/board/nap/nap_ext_event.c:
```diff
@@ -0,0 +1,97 @@
+  gps_time_t gpst = rx2gpstime(event_nap_time);
+  msg_ext_event_t msg;
+  msg.wn = gpst.wn;
+  msg.tow = gpst.tow * 1E3;
```
pedantry: we tend to use `1e3`

fnoble commented:
Looks good. One thought - we were trying to keep everything under `board/nap` as more of a library for interfacing with the NAP.

I would suggest making a function to read/write `NAP_REG_EXT_EVENT_TIME` in `nap_common.c` and then you can rename `src/board/nap/nap_ext_event.c` to `src/ext_event.c`. This keeps the firmware/OS code separate from the low level NAP read/write code.

If you wanted to split the `ext_event` functions out from `nap_common` then you could have a file called `src/board/nap/nap_ext_event.c` but I think it should just contain the register interface code.

swiftnav-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://zazu.ci.swiftnav.com/job/piksi_firmware_Pull_Requests/261/
Test PASSed.

swiftnav-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://zazu.ci.swiftnav.com/job/piksi_firmware_Pull_Requests/262/
Test PASSed.

henryhallam commented:
@fnoble OK, moved into main directory and rebased on master.  Bench testing looks fine, still showing the same occasional glitches.  I'll investigate those tomorrow but I think they're unrelated to this functionality and symptomatic of something in the PVT solver or elsewhere.
